include/dlaf/cuda_assert.h

```diff
@@ -0,0 +1,31 @@
+//
+// Distributed Linear Algebra with Future (DLAF)
+//
+// Copyright (c) 2018-2019, ETH Zurich
+// All rights reserved.
+//
+// Please, refer to the LICENSE file in the root directory.
+// SPDX-License-Identifier: BSD-3-Clause
+//
+
+#pragma once
+
+#include <cuda_runtime.h>
+#include <cstdio>
+#include <cstdlib>
+
+namespace dlaf {
+
+#ifdef DLAF_WITH_CUDA
+inline void cuda_assert(cudaError_t err, char const* file, int line) noexcept {
+  if (err != cudaSuccess) {
+    printf("[CUDA ERROR] %s:%d: '%s'\n", file, line, cudaGetErrorString(err));
+    std::abort();
+  }
+}
```

Comment:
What about using `DLAF_ASSERT` or something equivalent that it is not filtered by any flag (e.g. `DLAF_CHECK`)?

Using it we can centralize and make homogeneous the handling. It is possible to replicate the same trivial mechanism here, but it would be preferable to have it just in one place.

Moreover, `DLAF_ASSERT` uses `std::terminate` that allows to set a custom handler, that may be useful in our case where we may want to abort HPX + MPI + CUDA, depending on the context.

Reply:
> What about using DLAF_ASSERT or something equivalent that it is not filtered by any flag (e.g. DLAF_CHECK)?

`DLAF_ASSERT` and friends serve a different purpose, they allow us to check pre/postconditions and to assert statements within our functions and constructors. This is separate from error handling, it's what C++20 referred to as contracts. I made an unfortunate choice in naming the macro here. It should be DLAF_CUDA_CALL instead, mirroring the `MPI_CALL` macro (which I will also predix with DLAF_ in this PR). These macros handle errors reported by our dependencies (MPI and CUDA respectively), we have decided to just print what the errors are and abort but in principle we can make a different decision and continue executing if we know that we can recover from the error. In our case, we most likely can't, that is why we abort/terminate. HPX usually throws exceptions instead of using error codes to allow us to recover. 

As mentioned in this issue : https://github.com/eth-cscs/DLA-Future/issues/132 , there is a recent paper by Herb Sutter (one of the top C++ guys) regarding deterministic exceptions (http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p0709r4.pdf), the paper has the following section `4.2 Proposed cleanup: Don’t report logic errors using exceptions` (I recommend reading the section). This is what we want to do with our DLAF_ASSERT and friends, we don't want to throw exceptions or return error codes, we just want to inform the user that there is a non-recoverable bug. A lot of C++ libraries nowadays use exceptions for that but they miss out on some performance gains by not being able to mark their functions/constructors with `noexcept`.

Reply:
> Moreover, DLAF_ASSERT uses std::terminate that allows to set a custom handler, that may be useful in our case where we may want to abort HPX + MPI + CUDA, depending on the context.

Regarding `std::terminate()` vs `std::abort()`. HPX currently doesn't set a termination handler with `std::set_terminate()` except inside their unit tests. The default termination handler is just `std::abort()` (https://en.cppreference.com/w/cpp/error/terminate_handler). HPX does set a handler for `std::abort()` with `std::signal()` called `hpx::detail::on_abort()` in `src/runtime.cpp`. The handler just sets a global flag and calls `std::exit()`. HPX also sets a handler for both `std::exit()` and `std::quick_exit()` via `std::atexit()` and `std::at_quick_exit()` called `hpx::detail::on_exit()` in `src/runtime.cpp` that sets the same global flag.

I currently don't see any benefit of using `std::abort()` over `std::terminate()` so I will just use the latter as you suggested to be more uniform with other parts of the code. I am also not sure if we would want to use `std::set_terminate()` to set a handler yet. The only thing that annoys me a little bit with `std::terminate()` is that it prints a message if called directly as opposed to invoked via an uncaught exception.

Reply:
Found this gem that is worth a read on `std::terminate()` : https://akrzemi1.wordpress.com/2011/10/05/using-stdterminate/

Reply:
I completely agree with you about _"DLAF_ASSERT & Friends"_ and on the separated topic about checking CUDA and MPI calls (and also their names), in addition to the performance benefits we get by being able to specify `noexcept` (that IMHO we should use it more extensively).

About `std::terminate`, I think it is a nice placeholder for now and honestly I have not thought about the exit strategy for MPI+HPX. What I have in mind for the handler, it is that at least in MPI it would be nice to call `MPI_Abort` to signal other ranks that one of them failed.